### PR TITLE
fix: ensure response body has to be read fully and close correctly; a…

### LINKED ISSUE #2784: BUG: image-shim request registry link did not break, resulting in a large number of links
### Sealos Version

v4.1.4

### How to reproduce the bug?

查看cri-shim到registry的链接数量
image
image

### What is the expected behavior?

_No response_

### What do you see instead?

_No response_

### Operating environment

```markdown
- Sealos version:
- Docker version:
- Kubernetes version:
- Operating system:
- Runtime environment:
- Cluster size:
- Additional information:
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by fengxsong**
fix #2784

### CODE REVIEW

**Comment by fengxsong** [comment]
@bxy4543 do you have time to take a look?

File: pkg/utils/registry/registry.go [hunk 1]
```diff
@@ -18,12 +18,33 @@ package registry
...
+var (
+	sharedClients = make(map[*option]*registry.Registry)
+	lock          sync.RWMutex
+)
+
 func NewRegistry(url, username, password string, skipTLS bool) (*registry.Registry, error) {
+	key := &option{url, username, password, skipTLS}
+	lock.Lock()
+	defer lock.Unlock()
+	if v, ok := sharedClients[key]; ok {
```
Comment (bxy4543) [inline pkg/utils/registry/registry.go:41]: This doesn't always seem ok, you should replace the pointer type with a struct type, which will compare each property value of the struct.

**Comment by bxy4543** [comment]
> @bxy4543 do you have time to take a look?

This does not seem to be related to image-cri-shim and sealos.hub registry, how to solve this problem #2784

File: pkg/utils/registry/registry.go (hunk 1 above)
Comment (fengxsong) [reply pkg/utils/registry/registry.go:41]: > This doesn't always seem ok, you should replace the pointer type with a struct type, which will compare each property value of the struct.

make sense👍

**Review by bxy4543 — APPROVED** [review]

**Pull request merged**